Source/HTTP/Curl/CurlMulti.cpp

```diff
@@ -0,0 +1,157 @@
+#include "pch.h"
+#include "CurlMulti.h"
+#include "CurlProvider.h"
+
+namespace xbox
+{
+namespace http_client
+{
+
+// XCurl doesn't support curl_multi_timeout, so use a small, fixed delay between calls to curl_multi_perform
+#define PERFORM_DELAY_MS 50
+
+Result<HC_UNIQUE_PTR<CurlMulti>> CurlMulti::Initialize(XTaskQueuePortHandle workPort)
+{
+    assert(workPort);
+
+    http_stl_allocator<CurlMulti> a{};
+    HC_UNIQUE_PTR<CurlMulti> multi{ new (a.allocate(1)) CurlMulti };
+
+    multi->m_curlMultiHandle = curl_multi_init();
+    if (!multi->m_curlMultiHandle)
+    {
+        HC_TRACE_ERROR(HTTPCLIENT, "XCurlMulti::Initialize: curl_multi_init failed");
+        return E_FAIL;
+    }
+
+    RETURN_IF_FAILED(XTaskQueueCreateComposite(workPort, workPort, &multi->m_queue));
+
+    return Result<HC_UNIQUE_PTR<CurlMulti>>{ std::move(multi) };
+}
+
+CurlMulti::~CurlMulti()
+{
+    if (m_queue)
+    {
+        XTaskQueueTerminate(m_queue, true, nullptr, nullptr);
+        XTaskQueueCloseHandle(m_queue);
+    }
+
+    if (!m_easyRequests.empty())
+    {
+        HC_TRACE_WARNING(HTTPCLIENT, "XCurlMulti::~XCurlMulti: Failing all active requests.");
+        FailAllRequests(E_UNEXPECTED);
+    }
+
+    if (m_curlMultiHandle)
+    {
+        curl_multi_cleanup(m_curlMultiHandle);
+    }
+}
+
+HRESULT CurlMulti::AddRequest(HC_UNIQUE_PTR<CurlEasyRequest>&& easyRequest)
+{
+    std::unique_lock<std::mutex> lock{ m_mutex };
+
+    auto result = curl_multi_add_handle(m_curlMultiHandle, easyRequest->Handle());
```

Comment:
the lock probably solves any problem here, but would recommend doing this last after adding to your lists and such since you can start getting callbacks or in the worst case completion messages once this call is made when perform is called.

Reply:
with the lock, perform can't be called until the easy request is added to m_easyRequests and to the curl multi handle.  This order makes sense to me so that we don't have to worry about removing the request from m_easyRequests if curl_multi_add_handle_fails